### PR TITLE
Resolve ProcedureObserver errors on Xcode 9 beta 3

### PR DESCRIPTION
**Pull request by swiftlyfalling**
This is an alternative WIP to #753 that should maintain essentially all backwards-compatibility with existing code. 

**New functionality:**
- Catches incompatible Observers at run-time and provides helpful debug / logging output.

(So it theoretically could go into a ProcedureKit v4 point release, while the more significant / breaking changes in #753 are further-discussed.)

Thoughts and feedback appreciated.

### CODE REVIEW

File: Sources/ProcedureKit/AnyObserver.swift [hunk 1]
```diff
@@ -62,6 +62,103 @@ class AnyObserverBox<Base: ProcedureObserver>: AnyObserverBox_<Base.Procedure> {
...
+
+    private func typedProcedure(_ procedure: R, event: Event, logError: Bool = false) -> O? {
+        guard let typedProcedure = procedureTransformBlock(procedure) else {
+            procedure.log.warning(message: ("Observer will not receive event (\(event.string)). Unable to convert \(procedure) to the expected type \"\(String(describing: O.self))\""))
+            return nil
+        }
+        return typedProcedure
+    }
+
+    public func didAttach(to procedure: Procedure) {
```
Comment (danthorpe) [inline Sources/ProcedureKit/AnyObserver.swift]: Hmm, does the compiler not complain about these public functions, on an internal class?

**Comment by swiftlyfalling** [comment]
CI initially failed in the Stress Tests because of the rare issue fixed by #756

**Comment by danthorpe** [comment]
@swiftlyfalling - happy to merge?

**Comment by swiftlyfalling** [comment]
@danthorpe: Let's get #756 merged first, and then merge development back into this PR to verify that the Stress Test failure is resolved.

**Comment by swiftlyfalling** [comment]
Had one of those interesting CI test run failures (ProcedureKitNetwork) in which none of the tests run, and the test runner exits with error code 65. (Might have something to do with Fastlane.) Retrying the job.

**Pull request merged**